SMTChecker: Loop conditions should be analyzed as in loop context in BMC

blishko opened the pull request:
Value of loop condition can change between iterations of the loop.
Therefore, BMC should be in the "inside loop" state when analyzing loop conditions.

Previously, when visiting loop condition of a top-level loop, BMC engine did not consider itself being "inside loop", and therefore was checking if the condition is constant (always true or always false).
When the condition is always true in the first iteration and always false in the second iteration, BMC would try to report these two contradicting statements for the same source code expression, leading to an assertion violation in the `UniqueErrorReporter`.
Because the value of the condition can change between iteration, BMC should not try to check for constant condition in these cases.

As part of this PR the handling of the `ConstantCondition` verification target has been refactored (first commit).

Fixes #15847.

Commit:
SMTChecker: Refactor handling of ConstantCondition verification target

cameel commented on libsolidity/formal/BMC.cpp:
```diff
@@ -951,18 +935,35 @@ void BMC::checkVerificationTarget(BMCVerificationTarget& _target)
+
+	if (
+		!m_settings.targets.has(VerificationTargetType::ConstantCondition) ||
+		(m_currentContract && !shouldAnalyze(*m_currentContract))
```
Following the change from #15880, I guess we don't want this check to be bypassed simply because it's in an unselected contract, right?

```suggestion
		(m_currentContract && !m_currentContract->canBeDeployed())
```

blishko replied:
Correct! The code has been updated.

cameel replied:
Actually, I wonder why we even have to explicitly check `shouldEncode()` here. This seems to be the only such check in `BMC.cpp`. Why aren't there more, for other targets? Are we missing some?

cameel commented on libsolidity/formal/BMC.cpp:
```diff
@@ -951,18 +935,35 @@ void BMC::checkVerificationTarget(BMCVerificationTarget& _target)
+		return;
+
+	// Do not check for const-ness if this is a constant.
+	if (dynamic_cast<Literal const*>(&_condition))
```
It looks to me like this will not exclude constants (`bool constant C = true;`), just literals (`true`), right?

```suggestion
	// Do not check for const-ness if this is a literal.
	if (dynamic_cast<Literal const*>(&_condition))
```

blishko replied:
I am still confused about constants in Solidity. But I believe you are right and I've updated to comment accordingly.

cameel replied:
Well, yeah, there's a bit of fuzziness in our terminology here. There's one syntax element we call *constant* in Solidity: a state variable with `constant` applied to it. That's what comes to mind first when is see it in a comment that refers to an AST node. But of course there are a few other things that have constant values and it's not really wrong to call them constants. Literals are one of those.

In any case, this comment definitely is referring just to literals and not constants in general.

cameel commented on test/libsolidity/smtCheckerTests/loops/for_loop_ternary_operator_in_condition.sol:
I think this test would be better off named `top_level_for_loop_trivial_non_constant_condition.sol`. It would better explain why we're checking this, not just what we're testing.

blishko replied:
Actually, I don't think that captures the purpose of the test. I tried a different name.

cameel commented on Changelog.md:
```diff
@@ -22,6 +22,7 @@ Bugfixes:
  * Metadata: Fix custom cleanup sequence missing from metadata when other optimizer settings have default values.
  * SMTChecker: Fix internal compiler error when analyzing overflowing expressions or bitwise negation of unsigned types involving constants.
  * SMTChecker: Fix reporting on targets that are safe in the context of one contract but unsafe in the context of another contract.
+ * SMTChecker: Fix SMT logic error in BMC when loop condition is always true in one iteration and always false in the next one.
```
Looks like the ICE isn't the only thing this PR fixes. 0.8.28 reports `k == 0` as always false in this example:

```solidity
contract Test {
    function loop() public pure {
        uint k = 0;
        while (k == 0 ? true : false) {
            ++k;
        }
    }
}
```
```
Warning: BMC: Condition is always false.
 --> test.sol:4:16:
  |
4 |         while (k == 0 ? true : false) {
  |                ^^^^^^
Note: Callstack:
```

The version from your branch does not.

I'm comparing with 0.8.28 binary though so there's also a chance it's the result of another, already merged fix.

blishko replied:
You are right, this is the result of the current changes. I've updated the changelog entry, let me know what you think.

cameel commented on libsolidity/formal/BMC.cpp:
```diff
@@ -1195,10 +1194,6 @@ void BMC::checkBooleanNotConstant(
 	std::vector<SMTEncoder::CallStackEntry> const& _callStack
 )
 {
-	// Do not check for const-ness if this is a constant.
-	if (dynamic_cast<Literal const*>(&_condition))
-		return;
```
`checkBooleanNotConstant()`'s docstring needs to be updated. It still mentions that the function checks this.

blishko replied:
I update the docstring. Actually, coming back to this function, I refactored it further to only perform the satisfiability checks and moved reporting of the result to the only caller.

cameel commented on libsolidity/formal/BMC.cpp:
```diff
@@ -443,8 +445,10 @@ bool BMC::visit(ForStatement const& _node)
 		auto indices = copyVariableIndices();
 		if (_node.condition())
 		{
+			m_loopCheckpoints.emplace();
 			_node.condition()->accept(*this);
 			forCondition = expr(*_node.condition());
+			m_loopCheckpoints.pop();
```
You're adding/moving these in 3 different places, but you're only adding one new test case and not updating any existing ones. I suspect that we don't have coverage for all 3.

The example I posted under the changelog would probably cover one of them. What's the other one?

blishko replied:
I've added you example with `while` loop, which covers the change in `visit(WhileStatement)`.
The one test with for loop that I added originally actually exercises both changes in `visit(ForStatement)`, so no additional test is needed.

blishko replied:
To clarify, there are two places in `visit(ForStatement)` where the loop condition is visited:
Before each unrolling, and a final check when the loop no longer enters another iteration.
We needed to patch both places.

cameel replied:
Thanks for clarification. Looks good now.

Commit:
SMTChecker: Loop conditions should be analyzed as in loop context in BMC

Value of loop condition can change between iterations of the loop.
Therefore, BMC should be in the "inside loop" state when analyzing loop
conditions.

cameel commented on libsolidity/formal/BMC.h:
```diff
@@ -188,14 +192,17 @@ class BMC: public SMTEncoder
+	/// Checks whether the given boolean condition is either true or always false under given constraints.
+	/// Returns the results from the solver for these two checks.
+	ConstantExpressionCheckResult
+	checkBooleanNotConstant(smtutil::Expression const& _constraints, smtutil::Expression const& _condition);
```
The style with the return value on a separate line always looks confusing to me. I'd pretty much always prefer this:

```suggestion
	ConstantExpressionCheckResult checkBooleanNotConstant(
		smtutil::Expression const& _constraints,
		smtutil::Expression const& _condition
	);
```

blishko replied:
After having another look I agree that your version is more esthetically pleasing. I hope I got it right.

cameel reviewed (dismissed):
I have a small nitpick about style in one spot but other than that it looks good.
